app/src/test/java/echo/server/EchoServerTest.java

```diff
@@ -67,8 +58,51 @@ void should_sendMessage() throws IOException {
         .thenReturn("hello")
         .thenReturn(null);
 
-    echoServer.start();
+    echoServer.run();
 
     verify(mockSocketIO).sendMessage(mockServerConnection, "hello");
   }
+
+  @Test
+  @DisplayName("should return messages to multiple clients when more than one client sends a message")
+  public void should_sendMultipleClientsMessages() throws IOException, InterruptedException {
+    ExecutorService executorService = Executors.newCachedThreadPool();
+
+    // Start the server in a separate thread
+    executorService.execute(() -> {
+      try {
+        ServerRunner.main(null);
+      } catch (IOException e) {
+        e.printStackTrace();
+      }
+    });
+
+    // Wait for the server to start
+    Thread.sleep(1000);
+
+    int numClients = 5;
+    for (int i = 0; i < numClients; i++) {
+      // Start each client in a separate thread
+      int finalI = i;
+      executorService.execute(() -> {
+        try (Socket socket = new Socket("localhost", 49151);
+             PrintWriter out = new PrintWriter(socket.getOutputStream(), true);
+             BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()))) {
+
+          String message = "Hello from client " + finalI;
+          out.println(message);
+          String response = in.readLine();
+          assertEquals(message, response);
+        } catch (IOException e) {
+          e.printStackTrace();
+        }
+      });
+    }
+
+    // Wait for all clients to finish
+    executorService.shutdown();
+    Thread.sleep(5000);
+
+    executorService.shutdownNow();
+  }
```

Comment:
I'm not really 100% for using mocks "all" the time / dogmatic about it because mocks can be troublesome; if the interface for your dependency changes, the mock may not know it and won't fail for "the right reasons". If you spin up a real dependency, it will detect breaking changes. Concerns arise, for example, when the "real" dependency makes real IO interactions or real database calls because you don't want that in the test.